app/gui/view/graph-editor/src/lib.rs

```diff
@@ -1743,16 +1748,103 @@ impl GraphEditorModel {
         //   `find_free_place` looks for free place for the origin point, and we want to fit not
         //   only the point, but the whole node.
         let node_areas = nodes.values().map(|node| {
-            let position = node.position();
-            let left = position.x - x_gap - min_spacing;
-            let right = position.x + node.view.model.width() + x_gap;
-            let top = position.y + node::HEIGHT + y_gap;
-            let bottom = position.y - node::HEIGHT - y_gap;
+            let bounding_box = node.frp.bounding_box.value();
+            let left = bounding_box.left() - x_gap - min_spacing;
+            let right = bounding_box.right() + x_gap;
+            let top = bounding_box.top() + node::HEIGHT / 2.0 + y_gap;
+            let bottom = bounding_box.bottom() - node::HEIGHT / 2.0 - y_gap;
             OccupiedArea { x1: left, x2: right, y1: top, y2: bottom }
         });
         find_free_place(starting_from, direction, node_areas)
     }
 
+    /// Calculate a position for a new node, starting from the mouse position, but aligning to the
+    /// source node of the specified edge if the mouse position is close to the node.
+    ///
+    /// See [`new_node_position_aligned_if_close_to_node()`] for details on what "close to" means.
+    pub fn new_node_position_at_mouse_aligned_if_close_to_edge_source_node(
+        &self,
+        edge_id: EdgeId,
+        mouse_position: Vector2,
+    ) -> Vector2 {
+        let edge_source_node_id = self.edge_source_node_id(edge_id);
+        let edge_source_node = edge_source_node_id.and_then(|id| self.nodes.get_cloned_ref(&id));
+        self.new_node_position_aligned_if_close_to_node(mouse_position, edge_source_node)
+    }
+
+    /// Calculate a position for a new node, starting from the mouse position, but aligning to the
+    /// node closest to the mouse position if it is close enough.
+    ///
+    /// See [`new_node_position_aligned_if_close_to_node()`] for details on what is close enough.
+    pub fn new_node_position_at_mouse_aligned_to_close_nodes(
+        &self,
+        mouse_position: Vector2,
+    ) -> Vector2 {
+        let nearest_node = self.find_node_nearest_to_point(mouse_position);
+        self.new_node_position_aligned_if_close_to_node(mouse_position, nearest_node)
+    }
+
+    /// Finds a node nearest to the specified point.
+    ///
+    /// The distance between a point and a node is the distance between the point and the node's
+    /// bounding box.
+    fn find_node_nearest_to_point(&self, point: Vector2) -> Option<Node> {
```

Comment:
Why this name starts with `find` and other do not? For example `new_node_position_at_mouse_aligned_to_close_nodes` could also be named `find_new_node_position_at_mouse_aligned_to_close_nodes`, but it is not.